rl4co/envs/routing/pdp/env.py

```diff
@@ -209,3 +209,317 @@ def check_solution_validity(self, td, actions):
+        device: device to use.  Generally, no need to set as tensors are updated on the fly
+    """
+
+    name = "pdp_ruin_repair"
+
+    def __init__(
+        self,
+        generator: PDPGenerator = None,
+        generator_params: dict = {},
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        if generator is None:
+            generator = PDPGenerator(**generator_params)
+        self.generator = generator
+        self._make_spec(self.generator)
+
+    @staticmethod
+    def _step(td: TensorDict) -> TensorDict:
+        
+        # get state information from td
+        action = td["action"]
+        selected = action[:,0].view(-1,1)
+        first = action[:,1].view(-1,1)
+        second = action[:,2].view(-1,1)
+        solution = td["rec_current"]
+        solution_best = td["rec_best"]
+        locs = td["locs"]
+        cost_bsf = td["cost_bsf"]
+        action_record = td["action_record"]
+        bs, gs = solution.size()
+
+        # perform ruin and repair
+        next_rec = PDPRuinRepairEnv._insert_operator(solution, selected + 1, first, second)
+        new_obj = PDPRuinRepairEnv.get_costs(locs, next_rec)
+
+        # compute reward and update best-so-far solutions
+        now_bsf = torch.min(torch.cat((new_obj[:,None], cost_bsf[:, None]),-1),-1)[0]
+        reward = cost_bsf - now_bsf
+        index = reward > 0.0
+        solution_best[index] = next_rec[index].clone()
+        
+        # reset visited_time
+        visited_time = td["visited_time"] * 0
+        pre = torch.zeros((bs),device = visited_time.device).long()
+        arange = torch.arange(bs)
+        for i in range(gs):
+            current_nodes = next_rec[arange,pre]
+            visited_time[arange,current_nodes] = i+1
+            pre = next_rec[arange,pre]
+        visited_time = visited_time.long()
+        
+        # update action record
+        action_record[:,:-1] = action_record[:,1:]
+        action_record[:,-1] *= 0
+        action_record[torch.arange(bs), -1, action[:,0]] = 1
+
+        # Update step
+        td.update(
+            {
+                "cost_current": new_obj,
+                "cost_bsf": now_bsf,
+                "rec_current": next_rec,
+                "rec_best": solution_best,
+                "visited_time": visited_time,
+                "action_record": action_record,
+                "i": td["i"] + 1,
+                "reward": reward,
+                "done": td["done"],
+            }
+        )        
+        
+        return td
+
+    def _reset(self, td: Optional[TensorDict] = None, batch_size=None) -> TensorDict:
+        device = td.device
+
+        locs = torch.cat((td["depot"][:, None, :], td["locs"]), -2)
+        current_rec = self.generator._get_initial_solutions(locs)
+        obj = self.get_costs(locs, current_rec)
+        
+        # get index according to the solutions in the linked list data structure
+        bs = batch_size if isinstance(batch_size, int) else batch_size[0]
```

Comment:
[Minor] Getting `batch_size` is included in the `RL4COEnvBase`
https://github.com/ai4co/rl4co/blob/137097306a933db245b133d67f5c1a46bf1427e8/rl4co/envs/common/base.py#L141
And in the current version by default we consider the batch size is one dimension so don't worry~

Reply:
Sure, I changed here.